lib: tftp_client: Add TFTP PUT function

junqingzou opened the pull request:
Add tftp_put() API to support TFTP WRITE request.

Also add a callback for TFTP READ request so as to download a file with arbitary length.

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

mjico commented:
Also add a callback for TFTP READ request so as to download a file with arbitary length.

is this request in this code release？ it seems not include.

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -53,6 +53,61 @@ static size_t make_request(uint8_t *buf, int request,
 }
 
+/*
+ * Send an Error Message to the TFTP Server.
```
The comment seems wrong.

junqingzou replied:
Sorry C&P miss

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -247,7 +308,87 @@ int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
+	int ret;
+
+	if (client->user_buf == NULL || client->user_buf_size == 0) {
+		return -EFAULT;
```
I'd rather use EINVAL (invalid input parameters).

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -247,7 +308,87 @@ int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
+
+	sock = socket(server_addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("Failed to create UDP socket (IPv4): %d", errno);
```
Why IPv4? The server address seems to be application-provided, so it could be IPv6 as well I guess?

junqingzou replied:
C&P from tftp_get(), updated both places

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -53,6 +53,61 @@ static size_t make_request(uint8_t *buf, int request,
+			LOG_ERR("No more retransmits. Exiting");
+			return TFTPC_RETRIES_EXHAUSTED;
+		}
+		ret = send(sock, data_buffer, data_size + TFTP_HEADER_SIZE, 0);
```
`sendto()` should be used, there was no `connect()` on the socket.

junqingzou replied:
connect() has been done in send_request(), becoming connection-oriented UDP

rlubos replied:
Ah, ok indeed, quite hidden though.

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -53,6 +53,61 @@ static size_t make_request(uint8_t *buf, int request,
+
+		LOG_DBG("Receive: opcode %u, block no %u, size %d", opcode, blockno, ret);
+
+		if (opcode != ACK_OPCODE || blockno != block_no) {
```
I see a flaw in the current design - if the ACK is delayed for whatever reason, making us retransmit, but eventually arrives, the whole transmit produre will fail, as the routing will already expect the next block number. I think the routine should be more permissive here, i. e. allow receving ACKs with block number < that the current one, and simply ignore those.

junqingzou replied:
That might explains the problem I faced in upload streaming test.
Thanks for the suggestion!

rlubos reviewed (commented):
Please split the changes in `tftp_get()` into a separate commit, they're not related to the PUT implementation.

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -247,7 +308,87 @@ int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
+
+
+	do {
+		uint8_t send_buffer[TFTPC_MAX_BUF_SIZE];
```
This looks very dangerous to do in a library code as the buf size can be > 512 bytes, which means that you might be overwrite the stack. Please re-architecture your code to avoid this. One option is to let the caller provide a temp buffer for this function and use that.

junqingzou replied:
Based on the protocol, send and receive does not happen simultaneously.
Reuse the existing `tftpc_buffer` to send out data in 512-byte chunks.

jukkar replied:
That is indeed better but this library is not really a proper library because the buffer is shared between possible API callers. Perhaps this is not a real issue in the applications (that the API is not re-entrant), but perhaps this could be addressed at some point (fortunately the API is marked as experimental so should be easy to change if needed).

junqingzou replied:
The mutex could help a bit, but the real problem is that one same thread could lock a mutex multiple times.

junqingzou committed:
lib: tftp_client: Receive file with arbitrary size

Add a callback for READ request to download files with arbitrary length.
Define TFTP_BLOCK_SIZE in API header for application to allocate buffer.

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -53,6 +53,74 @@ static size_t make_request(uint8_t *buf, int request,
+
+	/* Receive ACK from the TFTP Server. */
+	count = 1;
+	do {
```
This is still not great, having to separate loops, impacts the retransmission logic. If for example a duplicate ACK is received, the function will no longer attempt to retransmit the current block.

What I think is needed, is an outer loop, dealing with retransmissions, and an inner loop within, with poll/recv, awaiting for ACKs and processing them. Only if the inner loop times-out, the outer loop should take another iteration, retransmitting the block.

junqingzou replied:
That is achievable, but maybe not reusing the buffer, that the sending buffer must be kept intact.

rlubos reviewed (commented):
Other than the one issue I've mentioned, it looks quite ok.

jukkar commented:
After looking the tftp client code, to me it looks like the API needs overhaul and it would probably be good to do that sooner than later. I am concerned that the API has shared resources like global receive buffer which can be easily corrupted if the application is not careful when calling the API.
Would it be a big problem to overhaul the API at this point? I am thinking something like application creating a context and passing that around in tftp api calls, instead of `int sock` and using global recv buffer.
WDYT

junqingzou commented:
> After looking the tftp client code, to me it looks like the API needs overhaul and it would probably be good to do that sooner than later. I am concerned that the API has shared resources like global receive buffer which can be easily corrupted if the application is not careful when calling the API. Would it be a big problem to overhaul the API at this point? I am thinking something like application creating a context and passing that around in tftp api calls, instead of `int sock` and using global recv buffer. WDYT

I agree that the code might need re-structured. Well the point of this PR to add the missing PUT and also enable large file downloading. Shall we update in two phases?

junqingzou commented:
Log of normal send and receive

[00:01:28.626,037] <dbg> tftp_client: send_request: Sending TFTP request 2 file test_put_02.txt
[00:01:28.953,033] <dbg> tftp_client: tftp_put: Receive: opcode 4, block no 0, size 4
[00:01:28.953,094] <dbg> tftp_client: send_data: Client send data: block no 1, size 512
[00:01:29.276,763] <dbg> tftp_client: send_data: Receive: opcode 4, block no 1, size 4
[00:01:29.276,763] <dbg> tftp_client: send_data: Client send data: block no 2, size 28
[00:01:29.581,756] <dbg> tftp_client: send_data: Receive: opcode 4, block no 2, size 4
[00:01:29.581,756] <dbg> tftp_client: tftp_put: 540 bytes sent.
[00:02:20.845,428] <dbg> tftp_client: send_request: Sending TFTP request 1 file test_put_02.txt
[00:02:21.393,402] <dbg> tftp_client: tftp_get: Received data: opcode 3, block no 1, size 516
[00:02:21.393,493] <dbg> tftp_client: send_ack: Client acking block number: 1
[00:02:21.679,779] <dbg> tftp_client: tftp_get: Received data: opcode 3, block no 2, size 32
[00:02:21.679,840] <dbg> tftp_client: send_ack: Client acking block number: 2

junqingzou committed:
lib: tftp_client: Add TFTP PUT function

Add tftp_put() API to support TFTP WRITE request.

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

rlubos commented:
>  I am concerned that the API has shared resources like global receive buffer which can be easily corrupted if the application is not careful when calling the API.

I see that the public API functions are protected with `tftpc_lock` mutex, shouldn't this be enough for a basic resource protection?

rlubos commented:
> I am thinking something like application creating a context and passing that around in tftp api calls, instead of int sock and using global recv buffer.
WDYT

True, this would follow similar pattern as we have in MQTT, where buffers for internal packet processing are allocated by the application, making the APIs more robust in terms of multi-threaded access. I think however this may extend a bit out of the scope of this PR, and  I agree with @junqingzou we could perhaps address this separately.

rlubos reviewed (dismissed):
Unless we disagree to introduce the tftp context separately, LGTM.

jukkar commented:
> I think however this may extend a bit out of the scope of this PR, and I agree with @junqingzou we could perhaps address this separately.

Ok, fair enough. It is just that adding more technical dept means that it will be more challenging to change the code later. I looked the code and it is rather simple still and changes would have been trivial.

jukkar reviewed (dismissed):
+1 in a hope that the code is changed and technical dept is removed in near future.

junqingzou commented:
> > I am concerned that the API has shared resources like global receive buffer which can be easily corrupted if the application is not careful when calling the API.
> 
> I see that the public API functions are protected with `tftpc_lock` mutex, shouldn't this be enough for a basic resource protection?

k_mutex_lock() allows to lock a mutex from same thread multiple times. If needed, I could change it to a binary semaphore.
Since we're talking about MQTT-style context, eventually neither mutex nor semaphore is required. But a new API like tftp_init() would be needed.
Another recommended design change should be that tftp_put() should support asynchrounous handling.

junqingzou commented:
Rebase to latest "main", trying to get rid of twister failure.

rlubos commented:
> k_mutex_lock() allows to lock a mutex from same thread multiple times. If needed, I could change it to a binary semaphore.

I don't see why such a change could be needed.

> Since we're talking about MQTT-style context, eventually neither mutex nor semaphore is required. But a new API like tftp_init() would be needed.

You'd still need a mutex, but as a part of the context.

junqingzou commented:
Yet another commit to catch up with Zephyr main.
I have also prepared a new version based on client context, ready to submit as next PR for your review.
